07.bowling_object/frame.rb

```diff
@@ -1,35 +1,63 @@
 # frozen_string_literal: true
 
 class Frame
-  def initialize(shot, status)
-    @first_shot = shot[0]
-    @second_shot = shot[1] ||= 0
-    @status = status
-  end
+  BONUS = {
+    double_strike: [3, 2],
+    strike: [2, 2],
+    strike_remainder: [2, 1],
+    spare: [2, 1],
+    normal: [1, 1]
+  }.freeze
 
-  BONUS = { double_strike: [3, 2],
-            strike: [2, 2],
-            strike_remainder: [2, 1],
-            spare: [2, 1],
-            normal: [1, 1] }.freeze
+  def initialize(shots, prev_status)
+    @shots = shots
+    @shots[1] ||= 0
+    @prev_status = prev_status
+  end
 
   def score
-    [@first_shot * BONUS[@status][0], @second_shot * BONUS[@status][1]].sum
+    2.times.sum { |i| @shots[i] * BONUS[@prev_status][i] }
   end
 
-  def next_status(frame_number)
-    if @first_shot == 10 && (@status == :double_strike || @status == :strike) && frame_number == 10
+  def current_status(frame_number)
+    if strike_remainder?(frame_number)
       :strike_remainder
-    elsif frame_number >= 10
+    elsif last_shot?(frame_number)
       :normal
-    elsif @first_shot == 10 && (@status == :double_strike || @status == :strike)
+    elsif double_strike?
       :double_strike
-    elsif @first_shot == 10
+    elsif strike?
       :strike
-    elsif [@first_shot, @second_shot].sum == 10
+    elsif spare?
       :spare
     else
       :normal
     end
   end
+
+  private
+
+  def last_shot?(frame_number)
+    frame_number >= 10
+  end
+
+  def prev_strike_or_double_strike?
+    @prev_status == :double_strike || @prev_status == :strike
+  end
+
+  def double_strike?
+    strike? && prev_strike_or_double_strike?
+  end
+
+  def strike?
+    @shots[0] == 10
+  end
+
+  def strike_remainder?(frame_number)
+    strike? && prev_strike_or_double_strike? && frame_number == 10
+  end
+
+  def spare?
+    @shots.sum == 10
```

Comment:
```suggestion
    !strike? && @shots.sum == 10
```

としておかないと、strikeでもspareと判定されそうです。
もちろん、今の実装であれば`@shots.sum == 10`だけでもちゃんと動くのですが、将来別の使い方がされるかもしれませんし、「自分でstrikeでないことを確認してからspareを呼ぶこと」という暗黙のルールを設けるより、いつ誰が呼んでもspareはspareと判定されるロジックになっている方が使いやすいクラスになります。

Reply:
言われてみれば[10, 0]は`@shots.sum == 10`の条件に当てはまりますが、明らかにスペアではないですね！
スペアの条件からstrikeの部分を除外するよう修正しました。[ec91f1c](https://github.com/tomonariha/ruby-practices/pull/12/commits/ec91f1c4fd8869b4bb3c238ec2ca1898f736862a)